storage/mysql/root_store_test.go

```diff
@@ -0,0 +1,324 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"encoding/base64"
+	"encoding/hex"
+	"flag"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/monologue/ctlog"
+	"github.com/google/monologue/storage/mysql/testdb"
+
+	_ "github.com/go-sql-driver/mysql" // Load MySQL driver
+)
+
+func mustCreateNewLog(url, name, b64PubKey string) *ctlog.Log {
+	l, err := ctlog.New(url, name, b64PubKey, 24*time.Hour, nil)
+	if err != nil {
+		glog.Fatalf("ctlog.New(%q, %q, %q) = _, %s", url, name, b64PubKey, err)
+	}
+	return l
+}
+
+var (
+	pilot = mustCreateNewLog("https://ct.googleapis.com/pilot", "pilot", "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEfahLEimAoz2t01p3uMziiLOl/fHTDM0YDOhBRuiBARsV4UvxG2LdNgoIGLrtCzWE0J5APC2em4JlvR8EEEFMoA==")
+)
+
+func mustParseCert(b64CertDER string, t *testing.T) *x509.Certificate {
+	t.Helper()
+	certDER, err := base64.StdEncoding.DecodeString(b64CertDER)
+	if err != nil {
+		t.Fatalf("Unexpected error while preparing testdata: %s", err)
+	}
+	cert, err := x509.ParseCertificate(certDER)
+	if err != nil {
+		t.Fatalf("Unexpected error while preparing testdata: %s", err)
+	}
+	return cert
+}
+
+func mustBytes32(in []byte, t *testing.T) [32]byte {
+	t.Helper()
+	if len(in) != 32 {
+		t.Fatalf("mustBytes32 expects input of length 32, got %d", len(in))
+	}
+	var out [32]byte
+	copy(out[:], in)
+	return out
+}
+
+func mustHexCode32(s string, t *testing.T) [32]byte {
+	t.Helper()
+	decoded, err := hex.DecodeString(s)
+	if err != nil {
+		t.Fatalf("Unexpected error while preparing testdata: %s", err)
+	}
+	return mustBytes32(decoded, t)
+}
+
+type rootEntry struct {
+	RootID    []byte
+	SHA256DER []byte
+}
+
+type setEntry struct {
+	RootSetID []byte
+	RootID    []byte
+}
+
+type obEntry struct {
+	LogName    string
+	RootSetID  []byte
+	ReceivedAt time.Time
+}
+
+func checkContents(ctx context.Context, t *testing.T, want []rootEntry, wantSets []setEntry, wantObservations []obEntry) {
+	t.Helper()
+
+	tx, err := testDB.BeginTx(ctx, nil /* opts */)
+	if err != nil {
+		t.Fatalf("failed to create transaction: %v", err)
+	}
+	defer tx.Commit()
+
+	// Roots
+	rows, err := tx.QueryContext(ctx, "SELECT Id, DER FROM Roots;")
```

Comment:
nit: You don't need the semi-colon at the end of statements passed to QueryContext().